Static Routes

Cydox opened the pull request:
This PR adds the ability to add static routes to a network. Discussion is in containers/netavark#678. The PR for support in netavark is containers/netavark#680


#### Does this PR introduce a user-facing change?




```release-note
This PR adds the --route flag and the no_default_route option in --opt. 

Example:
podman network create --subnet 10.10.10.0/24 --route 10.11.0.0/24,10.10.10.1,120 --opt no_default_route=1 testnet

This example creates a network with a static route to `10.11.0.0/24` via `10.10.10.1` 
and a route metric of `120`. In this example no default route will be added to 
containers in this network due to `no-default-route=1`. The route metric is optional. 
Multiple routes can be specified.
```

Luap99 commented:
/hold

You need to open a PR in c/common with the required changes there. And in either case we need to merge the implementation in netavark first merging anything in c/common or podman.

Luap99 commented on cmd/podman/networks/create.go:
```diff
@@ -52,6 +54,8 @@ func networkCreateFlags(cmd *cobra.Command) {
 	flags.IPSliceVar(&networkCreateOptions.Gateways, gatewayFlagName, nil, "IPv4 or IPv6 gateway for the subnet")
 	_ = cmd.RegisterFlagCompletionFunc(gatewayFlagName, completion.AutocompleteNone)
 
+	flags.BoolVar(&networkCreateOptions.NoAutoGateway, "no-auto-gateway", false, "do not automatically add gateways to subnets")
+
```
I am not a fan of this flag. Can you explain the use case more? Compared to e.g. --internal?
I really do not want to add more top level options to the Network type, if this is truly needed I think it is a much better fit via the generic options map.

Cydox replied:
On macvlan it does the exact same thing an `--internal`. I confirmed that all internal does on macvlan is prevent a default gateway. from being added automatically.

On bridge is where the difference comes in. Here, `--internal` also applies firewall rules that prevent external communication. So let's say you don't want a default route but add static routes to certain destinations. You can't achieve that with `--internal` as the firewall rules will block communication to those destinations.

Now that I write this, I think it might be possible to simply have the firewall rules allow communication to those static route destinations. Although that makes the meaning of `--internal` a little murky. In that case `--internal` would be internal except for those allowed packets to static route destinations.

Luap99 replied:
--internal does not add firewall rules to prevent external communication with bridge, we just set a sysctl to block routing on the bridge interface so it is impossible to reach ips outside the subnet.

I think having this as a use case is fine, but it should not be top level. Just put it in the existing `--opt` which just maps to a `[string]string` map, so something like `--opt no_auto_gateway=1` could be used.

We shouldn't change the meaning of internal at this point.

Cydox replied:
Sounds good

pjannesen replied:
A bridge network is a NAT router between the host and the pod/container. This NAT router uses the host route table. To make this work a default gateway for the bridge network is necessary.

On a macvlan network the gateway is explicitly defined with the --gateway option. We don't need no_auto_gateway for macvlan.

Combining bridge and macvlan in 1 pod/container is complex and requires rule-based routing. But this is complex and out of scope.

I only see multiple networks per pod/container with macvlan networks. I wouldn't allow a combination of macvlan and bridge networks. Therefore the no_auto_gateway flag is not needed.

Cydox replied:
The no_auto_gateway flag in this PR will be replaced by no_default_gateway. The relevant PR is containers/netavark#697





> A bridge network is a NAT router between the host and the pod/container. This NAT router uses the host route table. To make this work a default gateway for the bridge network is necessary.

For a bridge network a default route is not necessary if you have a relevant static route set up via the gateway.


> On a macvlan network the gateway is explicitly defined with the --gateway option. We don't need no_auto_gateway for macvlan.

On macvlan even without the --gateway option a gateway is added to the network config.

Cydox replied:
Edited the above comment, because I wrote no_default_gateway instead of no_default_route.

pjannesen replied:
> The no_auto_gateway flag in this PR will be replaced by no_default_route. The relevant PR is [containers/netavark#697](https://github.com/containers/netavark/pull/697)
> 
> > A bridge network is a NAT router between the host and the pod/container. This NAT router uses the host route table. To make this work a default gateway for the bridge network is necessary.
> 
> For a bridge network a default route is not necessary if you have a relevant static route set up via the gateway.

You can create static routes if you know the gateway address automatically defined by the bridge driver.
In a normal IP network, packets are allowed to follow different routes. Only the destination determines the route.
This does not apply to NAT routing. For connections that go over NAT, all packets must go through the same NAT router (because of connection tracking). This can be achieved with rule-based routing. However, this is out of scope I think.
Therefore I wouldn't recommend using bridge and macvlan in the pod/container.

> 
> > On a macvlan network the gateway is explicitly defined with the --gateway option. We don't need no_auto_gateway for macvlan.
> 
> On macvlan even without the --gateway option a gateway is added to the network config.

If the gateway address is not defined what address is then used? It is not wise not to make assumptions about what the gateway address might be (there is no standard). If this is already done by the macvlan driver, than this is unexpected behavior in my opinion.

Cydox replied:
To justify the feature in one (or two) sentence: If we give the user the ability to add routes to the container network namespace, then we should give him the ability to disable the otherwise automatically generated default route. That way he has full control over the routes in the network namespace if he desires.

Greetings from Delft, btw.

pjannesen replied:
I totally agree. multiple networks without static routes are almost useless. I don't like the no_default_path flag. But since podman assumes a gateway address without specifying one, I don't see any other solution.

We are researching podman for our internet gateways. A pod with 2 networks. 1 with a real internet ip number and 1 with an internal private ip address and containers with bind, nginx, postfix, asterisk, etc. So i hope this PR is in the next release :)

Podman also assumes a ip-range without specifying one (based on the subnet). This can result in duplicate IP addresses if you forget to specify an ip_address when creating the pod :-(.
It would be nice to have 'none' for --ip-range and --gateway.

I also miss a standard way to create firewall rules (nftable/iptables) for a pod.

Cydox commented:
> You need to open a PR in c/common with the required changes there. 

It was a little difficult to deal with the stuff that's in common. Am I doing this workflow the right way? I made a fork of common with the changes and added a `replace` in the `go.mod` here. So when I open a PR in common and that is merged, this PR requires a change to the `go.mod` before merging. Is there a better way around this?

Luap99 commented:
> > You need to open a PR in c/common with the required changes there.
> 
> It was a little difficult to deal with the stuff that's in common. Am I doing this workflow the right way? I made a fork of common with the changes and added a `replace` in the `go.mod` here. So when I open a PR in common and that is merged, this PR requires a change to the `go.mod` before merging. Is there a better way around this?

Yes this is the correct flow to test it here in podman. For actual merging you need to open the c/common changes as PR there. When it is merged there you just update the dependency here instead of the replace.

Cydox commented:
Made the changes. So now the way to prevent a default gateway if with `--opt no-auto-gateway=1`. Edited the "Does this PR introduce a user-facing change?" section accordingly. Also opened a PR in common with the needed changes.

Looks like copr build don't work when there are merge conflicts. Won't rebase at this point since at least the go.mod needs to updated anyways.

Cydox commented:
All tests passing, will just need a rebase once the PR in netavark and the PR in common are in to make rpm-builds work.

Cydox commented:
Rebased this commit and incorporated the changes needed to switch to `no_default_route` instead of `no_auto_gateway`, as well as the other changes from containers/common#1440

Cydox commented:
CI failed because some mirror seems to be down. Trying again.
/test all

Cydox commented:
The one failed test appears to be a gateway timeout from quay.io

Can't quite make sense of the other failure.

Cydox commented:
Updated this PR with all the latest changes in containers/common#1440 to test everything together.

TomSweeneyRedHat commented on docs/source/markdown/podman-network-create.1.md:
```diff
@@ -95,10 +95,12 @@ Set metadata for a network (e.g., --label mykey=value).
 
 - `mtu`: Sets the Maximum Transmission Unit (MTU) and takes an integer value.
 - `metric` Sets the Route Metric for the default route created in every container joined to this network. Accepts a positive integer value. Can only be used with the Netavark network backend.
+- `no_default_route`: If set to 1, podman will not automatically add a default route to subnets. Routes can still be added
```
```suggestion
- `no_default_route`: If set to 1, Podman will not automatically add a default route to subnets. Routes can still be added
```

TomSweeneyRedHat commented on docs/source/markdown/podman-network-create.1.md:
```diff
@@ -114,6 +116,10 @@ The `macvlan` and `ipvlan` driver support the following options:
 
+#### **--route**=*route*
+
+A static route in the format \<destination in CIDR notaion>,\<gateway>,\<route metric (optional)>. This route will be added to every container in this network. Only available with the netavark backend. Can be specified multiple times if more than one static route is desired.
```
```suggestion
A static route in the format \<destination in CIDR notation>,\<gateway>,\<route metric (optional)>. This route will be added to every container in this network. Only available with the netavark backend. It can be specified multiple times if more than one static route is desired.
```

TomSweeneyRedHat replied:
and do you need to backslash the `<`, but not the `>` symbols?

Cydox replied:
No those "\\" are a typo (not sure how that happened...). I'll fix them.

Cydox replied:
It's not xml. Here's an example route:

`--route 10.0.10.0/24,10.0.0.1,100`

Not sure if my notation here is the best way to describe that.

TomSweeneyRedHat commented on go.mod:
```diff
@@ -2,6 +2,8 @@ module github.com/containers/podman/v4
 
 go 1.18
 
+replace github.com/containers/common v0.53.1-0.20230506101404-3e93a76d461c => github.com/Cydox/common v0.0.0-20230523224052-80eac653d332
```
You'll need to get your changes from your fork into the upstream branch and vendor from there before merging this.

TomSweeneyRedHat replied:
I would just revendor in your sandbox, using the latest upstream c/common with your change in it, then repush this PR.

Cydox commented:
I vendored in the latest containers/common with containers/common#1440. Obviously this pulls in a lot of unrelated changes from c/common, so I'm not sure if this should be split into a separate PR.

 I split all the vendoring stuff into a separate commit (let me know if you prefer having this squashed)

[Now the CI](https://cirrus-ci.com/build/5307174964953088) is failing because of the bloat-check. Size of bin/podman increases by 171 kB. In the [previous CI run](https://cirrus-ci.com/build/5895494921093120) were I only vendored in a branch of common with the changes needed for this PR, the size increase was only 1.9 kB. So the lion-share of this increase is from the unrelated upstream changes in c/common.

@Luap99 How do we proceed?

Cydox commented:
Also, [sys podman debian-12 rootless host boltdb](https://cirrus-ci.com/task/6548896986431488?logs=main#L655) failed, but I believe this might be a weird random failure. I think I've seen a similar failure before that disappeared without any changes.

Luap99 commented:
> Also, [sys podman debian-12 rootless host boltdb](https://cirrus-ci.com/task/6548896986431488?logs=main#L655) failed, but I believe this might be a weird random failure. I think I've seen a similar failure before that disappeared without any changes.

Yes vendoring is broken right now: https://github.com/containers/podman/pull/18510. You have to wait until somebody from our team fixes that.

Cydox commented:
Thanks @Luap99 

Saw that it was cherry picked into that PR.

It should all be working as expected, I've been running the rpm from this PR on a local  fedora coreos server and it's all working as expected for my use-case.

Will this correctly make its way into the release notes with this PR being closed?
